src/test/java/com/example/shopapp/services/ClientServiceTest.java

```diff
@@ -0,0 +1,190 @@
+                mockProduct(3L, 5.5),
+                mockProduct(4L, 4.5),
+                mockProduct(5L, 5.0)))
+                .when(productRepo).findAllById(productIdsList);
+    }
+
+    @Test
+    void testGetOrderProductsByIds() {
+        var clientWrapper = clientWrapperSetUp();
+        var orderProductIdsList = orderProductsIdListSetUp(clientWrapper);
+
+        var orderProductsDTO = tested.getOrderProductsByIds(clientWrapper);
+
+        assertEquals(6, orderProductsDTO.size());
+        verify(productRepo).findAllById(orderProductIdsList);
+        verify(productMapper, times(6)).toDto(any(Product.class));
+        assertEqualsIds(orderProductsDTO);
+
+    }
+
+    @Test
+    void testAddClientAndOrder() {
+        var clientWrapper = clientWrapperSetUp();
+        var client = toEntity(clientWrapper.getClient());
+        stubbingSetUp(client);
+        Order order = orderSetUp(clientWrapper, client);
+
+        verify(productMapper, times(12)).toEntity(any(ProductDTO.class));
+        verify(clientRepo, atLeast(0)).save(client);
+        verify(orderRepo, atLeast(0)).save(order);
+        assertEqualsIds(order);
+        assertEqualsPrices(order);
+    }
+
+    private Product mockProduct(Long id, Double price) {
+
+        var product = mock(Product.class);
+        doReturn(id).when(product).getProductId();
+        doReturn(price).when(product).getPrice();
+        return product;
+    }
+
+    private ClientWrapper clientWrapperSetUp() {
+        var clientWrapper = new ClientWrapper();
+        var clientDTO = new ClientDTO(1L, "Michael", "Jordan", "Warsaw",
+                "Prosta", "50-500", "Poland");
+        var theOrderProductIds = "1,4,3,2,5,5";
+        clientWrapper.setOrderProductIds(theOrderProductIds);
+        clientWrapper.setClient(clientDTO);
+
+        return clientWrapper;
+    }
+
+    private void stubbingSetUp(Client client) {
+        when(clientMapper.toEntity(any(ClientDTO.class)))
+                .thenReturn(Client.builder().build());
+        when(clientRepo.save(any(Client.class))).thenReturn(client);
+        when(productMapper.toEntity(any(ProductDTO.class))).thenAnswer(invocationOnMock -> {
+            ProductDTO productDTO = invocationOnMock.getArgument(0);
+            return Product.builder().productId(productDTO.getProductId()).price(productDTO.getPrice()).build();
+        });
+    }
+
+    private Order orderSetUp(ClientWrapper clientWrapper, Client client) {
+        var productsDTOList = tested.getOrderProductsByIds(clientWrapper);
+        var productsList = productsDTOList.stream()
+                .map(productMapper::toEntity)
+                .toList();
+        var totalPrice = getTotalPrice(productsList);
+        var order = new Order(1L, totalPrice, productsList, client);
+        when((orderRepo).save(any(Order.class))).thenReturn(order);
+        tested.addClientAndOrder(clientWrapper);
+
+        return order;
+    }
+
+    private void assertEqualsIds(Order order) {
```

Comment:
Transform those 3 to var-args methods like this
`private void assertEqualsIds(Order order, Long... expectedProductIds)`
and make the assertions in a loop